L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc

```diff
@@ -0,0 +1,1094 @@
+    }
+  }
+}
+
+// ------------ method called to produce the data  ------------
+void L1TCorrelatorLayer1Producer::produce(edm::Event &iEvent, const edm::EventSetup &iSetup) {
+  // clear the regions also at the beginning, in case one event didn't complete but the job continues on
+  initEvent(iEvent);
+
+  /// ------ READ TRACKS ----
+  if (hasTracks_) {
+    edm::Handle<l1t::PFTrackCollection> htracks;
+    iEvent.getByToken(tkCands_, htracks);
+    const auto &tracks = *htracks;
+    for (unsigned int itk = 0, ntk = tracks.size(); itk < ntk; ++itk) {
+      const auto &tk = tracks[itk];
+      // adding objects to PF
+      if (debugR_ > 0 && deltaR(tk.eta(), tk.phi(), debugEta_, debugPhi_) > debugR_)
+        continue;
+      if (tk.pt() > trkPt_) {
+        addTrack(tk, l1t::PFTrackRef(htracks, itk));
+      }
+    }
+  }
+
+  /// ------ READ MUONS ----
+  edm::Handle<l1t::SAMuonCollection> muons;
+  iEvent.getByToken(muCands_, muons);
+  for (unsigned int i = 0, n = muons->size(); i < n; ++i) {
+    const l1t::SAMuon &mu = (*muons)[i];
+    if (debugR_ > 0 && deltaR(mu.eta(), mu.phi(), debugEta_, debugPhi_) > debugR_)
+      continue;
+    addMuon(mu, l1t::PFCandidate::MuonRef(muons, i));
+  }
+  // ------ READ CALOS -----
+  edm::Handle<l1t::PFClusterCollection> caloHandle;
+  for (const auto &tag : emCands_) {
+    iEvent.getByToken(tag, caloHandle);
+    const auto &calos = *caloHandle;
+    for (unsigned int ic = 0, nc = calos.size(); ic < nc; ++ic) {
+      const auto &calo = calos[ic];
+      if (debugR_ > 0 && deltaR(calo.eta(), calo.phi(), debugEta_, debugPhi_) > debugR_)
+        continue;
+      if (calo.pt() > emPtCut_)
+        addEmCalo(calo, l1t::PFClusterRef(caloHandle, ic));
+    }
+  }
+  for (const auto &tag : hadCands_) {
+    iEvent.getByToken(tag, caloHandle);
+    const auto &calos = *caloHandle;
+    for (unsigned int ic = 0, nc = calos.size(); ic < nc; ++ic) {
+      const auto &calo = calos[ic];
+      if (debugR_ > 0 && deltaR(calo.eta(), calo.phi(), debugEta_, debugPhi_) > debugR_)
+        continue;
+      if (calo.pt() > hadPtCut_)
+        addHadCalo(calo, l1t::PFClusterRef(caloHandle, ic));
+    }
+  }
+
+  regionizer_->run(event_.decoded, event_.pfinputs);
+
+  // First, get a copy of the discretized and corrected inputs, and write them out
+  iEvent.put(fetchEmCalo(), "EmCalo");
+  iEvent.put(fetchHadCalo(), "Calo");
+  iEvent.put(fetchTracks(), "TK");
+
+  // Then do the vertexing, and save it out
+  std::vector<float> z0s;
+  std::vector<std::pair<float, float>> ptsums;
+  float z0 = 0;
+  double ptsum = 0;
+  l1t::VertexWord pvwd;
+  // FIXME: collections seem to be already sorted
+  if (emuTkVtx_) {
+    edm::Handle<std::vector<l1t::VertexWord>> vtxEmuHandle;
+    iEvent.getByToken(tkVtxEmu_, vtxEmuHandle);
+    for (const auto &vtx : *vtxEmuHandle) {
+      ptsums.push_back(std::pair<float, float>(vtx.pt(), vtx.z0()));
```

Comment:
May be useful a reserve for `ptsums` with `vtxEmuHandle` above (below in all `else`s).